### ISSUE TITLE
Use XLA to Tacotron2 is slower than without XLA

### ISSUE THREAD

**Issue by lee-bin** [tensorflow/tensorflow#30791]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):centos
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):binary
- TensorFlow version (use command below):v1.12.0-0-ga6d8ffae09 1.12.0
- Python version:3.6.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:V9.0.176 / 7
- GPU model and memory: P40

**Describe the current behavior**
Before using XLA, the [Tacotron 2](https://github.com/Rayhane-mamah/Tacotron-2) runs about 0.5s, however after using XLA, it increases to about 0.7s which is much slower.
```python
config = tf.ConfigProto()
config.graph_options.optimizer_options.global_jit_level = tf.OptimizerOptions.ON_1              
self.session = tf.Session(config=config) 
```
I use XLA like above. Any thoughts would be appreciated. Thanks.

**Comment by tpopp** [comment]
Hi, sorry to hear about the performance problems.

The checklist that comes to mind immediately is:
- When testing, make sure to take the graph and run it multiple times then take the lowest number. XLA is a jit, so sometimes people compare the number on the first run, and that includes the compile time. Averaging multiple runs will also have the same problem for small times like this also.
- It's also possible that we compile many times if this model has varying shape sizes. If you try setting the environment variable `TF_CPP_VMODULE=xla_compilation_cache=2` and are not seeing messages saying there are cache hits, then it's still a compile time issue due to varying shapes.
- A variant of the compile problem is that we modify the graph to run XLA, which can hurt performance when XLA cannot provide benefits when it compiles the modified graph. This would be the case if you set the environment variable `TF_XLA_FLAGS=--tf_xla_always_defer_compilation=true` and get the 0.7 number still. This indicates that the graph rewriting slows down the model, and that XLA is not providing benefits on this modified graph.

If these compile related issues are not the problem, then it sounds like we're actually producing slow code in some way. Let's start with checking these points then go from there?

**Comment by lee-bin** [comment]
Thanks for your detailed answer. @tpopp 
- I skipped the first run and averaging 1000 runs after the first run. Is averaging multiple runs excluding the first run OK?
- I tried setting `TF_CPP_VMODULE=xla_compilation_cache=2`, nothing was printed. I used the Python API as the code is written in Python, does this environment variable work with Python API?
- I tried setting `TF_XLA_FLAGS=--tf_xla_always_defer_compilation=true`, the averaging time of 1000 runs was about the same.
- One more situation is that I tried to run the [tutorial demo](https://www.tensorflow.org/xla/jit#tutorial), run with XLA still lags behind run without XLA.
This is run without XLA.
![image](https://user-images.githubusercontent.com/29224933/61631025-cd9d2300-acbb-11e9-972d-583e9e3affe3.png)
This is run with XLA.
![image](https://user-images.githubusercontent.com/29224933/61630884-7ac36b80-acbb-11e9-9c0f-fcb546bcf267.png)

**Comment by tpopp** [comment]
Averaging the 1000 runs is fine as long as we aren't recompiling based on different steps. And the tutorial demo not being faster is probably reasonable. It's a small piece of code which gives us less opportunity to optimize.

I also realize that I wasn't fully clear on my directions, so I want to confirm that both `TF_CPP_VMODULE=xla_compilation_cache=2` and `TF_XLA_FLAGS=--tf_xla_always_defer_compilation=true` should only have an effect (printing and possibly changing timing respectively) when using XLA, and "always_defer_compilation" will cause the TF_CPP_VMODULE to not have an effect, so they should not be used together.

Regarding the VMODULE issue, can you see if running the command below in a shell creates additional logs with the line "xla_compilation_cache" in them?

```
TF_CPP_VMODULE=xla_compilation_cache=2 TF_XLA_FLAGS='--tf_xla_auto_jit=1 --tf_xla_cpu_global_jit --tf_xla_min_cluster_size=1' python -c "import tensorflow as tf; sess = tf.Session(); print(tf.reduce_sum(tf.random.normal([1000, 1000]) + 1).eval(session=sess))"
```

I would recommend the next 2 steps:
Running the model with the environment variables will dump many outputs to /tmp/tacotron/. If you're able to share the output in file "/tmp/tacotron//after_phase_30_PartiallyDeclusterPass[_*].pbtxt" we can try to see where the problem might be:
TF_DUMP_GRAPH_PREFIX=/tmp/tacotron/ TF_CPP_VMODULE=optimization_registry=2

The best thing would be if you can obtain a trace like you did with the tutorial, but that might be a lot of work. The most likely outcome of this research will be that this model has pieces of Tensorflow functionality that we can't support in XLA, so we're causing some additional overhead through control flow, then providing no speedups.

**Comment by lee-bin** [comment]
Thanks again for your detailed analysis.

I'm doing inference instead of training, and the input is the same for 1000 runs, but as there is dropout during inference, the length of output may be different. I don't know if this could cause recompiling.

I use `TF_CPP_VMODULE=xla_compilation_cache=2` and `TF_XLA_FLAGS=--tf_xla_always_defer_compilation=true` separately.

I tried running the command below, but found no prints with "xla_compilation_cache" in them.
```
$ TF_CPP_VMODULE=xla_compilation_cache=2 TF_XLA_FLAGS='--tf_xla_auto_jit=1 --tf_xla_cpu_global_jit --tf_xla_min_cluster_size=1' python -c "import tensorflow as tf; sess = tf.Session(); print(tf.reduce_sum(tf.random.normal([1000, 1000]) + 1).eval(session=sess))"
2019-07-24 11:39:14.339840: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-07-24 11:39:14.775164: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1432] Found device 0 with properties: 
name: Tesla P40 major: 6 minor: 1 memoryClockRate(GHz): 1.531
pciBusID: 0000:83:00.0
totalMemory: 22.38GiB freeMemory: 22.21GiB
2019-07-24 11:39:14.775238: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1511] Adding visible gpu devices: 0
2019-07-24 11:39:15.137972: I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-07-24 11:39:15.138047: I tensorflow/core/common_runtime/gpu/gpu_device.cc:988]      0 
2019-07-24 11:39:15.138063: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 0:   N 
2019-07-24 11:39:15.138596: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 21544 MB memory) -> physical GPU (device: 0, name: Tesla P40, pci bus id: 0000:83:00.0, compute capability: 6.1)
2019-07-24 11:39:15.490174: I tensorflow/compiler/xla/service/service.cc:149] XLA service 0x7f0b6c002210 executing computations on platform CUDA. Devices:
2019-07-24 11:39:15.490259: I tensorflow/compiler/xla/service/service.cc:157]   StreamExecutor device (0): Tesla P40, Compute Capability 6.1
2019-07-24 11:39:15.633148: W tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc:402] *** WARNING *** You are using ptxas 9.0.176, which is older than 9.2.88. ptxas 9.x before 9.2.88 is known to miscompile XLA code, leading to incorrect results or invalid-address errors.

You do not need to update to CUDA 9.2.88; cherry-picking the ptxas binary is sufficient.
1001091.1
```

I can't find anything under /tmp/tacotron/ after running the Tacotron model with environment variables `TF_DUMP_GRAPH_PREFIX=/tmp/tacotron/ TF_CPP_VMODULE=optimization_registry=2`, so I recorded the timeline [with](https://drive.google.com/open?id=1J7yyX480hc0JrVzSRTfiQ-sxPQ5u47Mn) and [without](https://drive.google.com/open?id=10ciVqnzM6DQZtS_fRE1x4lH4XGpnCwbY) XLA. Maybe you can find something useful in timeline, thanks.

**Comment by tpopp** [comment]
Just a warning, I am without my work computer until next Wednesday or Tuesday. I will look at this at that time, but I won't get a chance to before.

That's strange that the TF_CPP_VMODULE is not working for you, but we will try to work around it.

**Comment by lee-bin** [comment]
No problems, thanks!

**Comment by tpopp** [comment]
So in this case, it does not seem like XLA is a problem. To be specific, there are two parts to XLA used by Tensorflow. There is a "tf2xla bridge" that changes the graph to be compatible with XLA, then XLA itself. The latter is not clearly a problem based on your timelines. The former (the bridge) does seem to be an issue though. We would hope for the timeline to be primarily a single XlaCompile and then a very long XlaRun operation, and almost nothing else.

I figured out the TF_CPP_VMODULE issue. I thought this feature was around longer, but it is not in tf 1.12. If you can upgrade tensorflow to tf 1.14 is the simplest solution to make the commands work out. I will post in a couple hours the commands to run for tf 1.12. There will be less debug information but hopefully still enough.

**Comment by tpopp** [comment]
To get additional information with your version of Tensorflow, you can add:
TF_CPP_MIN_VLOG_LEVEL=2 TF_XLA_FLAGS="--tf_dump_graph_prefix=/tmp/tacotron/ --tf_xla_clustering_debug". You might want to redirect output to a file when running this because much other unneeded information will be printed that will flood your screen.  The files logged out with this can be analyzed to figure out why we aren't running XLA on a larger amount of the graph.

**Comment by lee-bin** [comment]
Thanks again for your patience.
I ran the command with the environment variables above and recorded outputs to [the file](https://drive.google.com/open?id=17ncyy2Xl0jRJBxdMPR8oojqdV0OT6P8V). It's a very big file. Maybe you can help figure out what happened. :)

**Comment by lee-bin** [comment]
Any thoughts? @tpopp

**Comment by sanjoy** [comment]
@tpopp is out of office this week so he won't be able to reply before next week.

Let me know if this is urgent and I can try to take a look.

**Comment by lee-bin** [comment]
Thanks, @sanjoy, next week is fine.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
